File: backend/routes/interaction_routes.py
"""Like / save + saved-listings routes."""
from __future__ import annotations
from fastapi import APIRouter, Depends

from middleware.auth_middleware import require_auth
from services import interaction_service

router = APIRouter(prefix="/v1", tags=["interactions"])


@router.post("/listings/{listing_id}/like")
async def toggle_like(listing_id: str, user=Depends(require_auth)):
    return await interaction_service.toggle(user.id, listing_id, "like")


@router.post("/listings/{listing_id}/save")
async def toggle_save(listing_id: str, user=Depends(require_auth)):
    return await interaction_service.toggle(user.id, listing_id, "save")


@router.get("/interactions/me/saved")
async def my_saved(user=Depends(require_auth)):
    items = await interaction_service.my_listings_by_type(user.id, "save")
    return {"items": items}


@router.get("/interactions/me/liked")
async def my_liked(user=Depends(require_auth)):
    items = await interaction_service.my_listings_by_type(user.id, "like")
    return {"items": items}

File: backend/src/routes/interaction.routes.js
const express = require('express');
const { requireAuth } = require('../middleware/auth.middleware');
const interactionService = require('../services/interaction.service');
const { catchAsync } = require('../utils/helpers');

const router = express.Router();

router.post('/listings/:listing_id/like', requireAuth, catchAsync(async (req, res) => {
  const result = await interactionService.toggle(req.user._id.toString(), req.params.listing_id, 'like');
  res.json(result);
}));

router.post('/listings/:listing_id/save', requireAuth, catchAsync(async (req, res) => {
  const result = await interactionService.toggle(req.user._id.toString(), req.params.listing_id, 'save');
  res.json(result);
}));

router.get('/interactions/me/saved', requireAuth, catchAsync(async (req, res) => {
  const items = await interactionService.myListingsByType(req.user._id.toString(), 'save');
  res.json({ items });
}));

router.get('/interactions/me/liked', requireAuth, catchAsync(async (req, res) => {
  const items = await interactionService.myListingsByType(req.user._id.toString(), 'like');
  res.json({ items });
}));

module.exports = router;
